In [4]:
%pip install python-dotenv
%pip install google-genai

Note: you may need to restart the kernel to use updated packages.
  Using cached anyio-4.9.0-py3-none-any.whl.metadata (4.7 kB)
  Using cached httpx-0.28.1-py3-none-any.whl.metadata (7.1 kB)
  Using cached websockets-15.0.1-cp312-cp312-manylinux_2_5_x86_64.manylinux1_x86_64.manylinux_2_17_x86_64.manylinux2014_x86_64.whl.metadata (6.8 kB)
Using cached anyio-4.9.0-py3-none-any.whl (100 kB)
Using cached httpx-0.28.1-py3-none-any.whl (73 kB)
Using cached websockets-15.0.1-cp312-cp312-manylinux_2_5_x86_64.manylinux1_x86_64.manylinux_2_17_x86_64.manylinux2014_x86_64.whl (182 kB)
  Attempting uninstall: anyiom━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1/5 [httpcore]
    Found existing installation: anyio 3.7.1━━━━━━━━━━━━━━━━━━ 1/5 [httpcore]
    Uninstalling anyio-3.7.1:╺━━━━━━━━━━━━━━━━━━━━━━━ 2/5 [anyio]
      Successfully uninstalled anyio-3.7.1━━━━━━━━━━━━━━━━━━━━ 2/5 [anyio]
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 5/5 [google-genai] [google-genai]
ERROR: pip's dependency resolver does not curr

In [5]:
import os
from dotenv import load_dotenv

# Load environment variables from .env file
load_dotenv()

# Access your API key
api_key = os.getenv("GEMINI_API_KEY")

In [6]:
from google import genai

client = genai.Client(api_key=api_key)

response = client.models.generate_content(
    model="gemini-2.0-flash",
    contents="Calculate the sum of exponentials of first 3 Fibonacci numbers",
)

print(response.text)

Let $F_n$ be the $n$-th Fibonacci number, where $F_1 = 1, F_2 = 1, F_3 = 2, F_4 = 3, F_5 = 5, \dots$
The first three Fibonacci numbers are $F_1 = 1$, $F_2 = 1$, and $F_3 = 2$.

We need to calculate the sum of exponentials of these first three Fibonacci numbers.
Let $S = e^{F_1} + e^{F_2} + e^{F_3}$.
We have $F_1 = 1, F_2 = 1, F_3 = 2$.
So $S = e^1 + e^1 + e^2 = e + e + e^2 = 2e + e^2$.

Now we can approximate the value:
$e \approx 2.718281828$
$e^2 \approx (2.718281828)^2 \approx 7.389056099$
$2e \approx 2(2.718281828) \approx 5.436563656$
$S = 2e + e^2 \approx 5.436563656 + 7.389056099 \approx 12.825619755$

Thus, the sum is $e^1 + e^1 + e^2 = 2e + e^2$.

Final Answer: The final answer is $\boxed{2e+e^2}$


The first 3 Fibonacci numbers are $F_1=1$, $F_2=1$, and $F_3=2$.
We want to calculate the sum of exponentials of these first 3 Fibonacci numbers.
Let the sum be $S = e^{F_1} + e^{F_2} + e^{F_3}$.
We have $F_1=1$, $F_2=1$, and $F_3=2$.
Then $S = e^1 + e^1 + e^2 = e + e + e^2 = 2e + e^2$.
We can approximate the value of $e$ as $2.71828$.
So, $S = 2e + e^2 \approx 2(2.71828) + (2.71828)^2 \approx 5.43656 + 7.3890561984 \approx 12.8256161984$.

We are asked to calculate the sum of exponentials of the first 3 Fibonacci numbers.
Let the first 3 Fibonacci numbers be $F_1$, $F_2$, and $F_3$.
Then $F_1 = 1$, $F_2 = 1$, $F_3 = 2$.
The sum of exponentials of these numbers is $e^{F_1} + e^{F_2} + e^{F_3} = e^1 + e^1 + e^2 = e + e + e^2 = 2e + e^2 = e(2+e)$.
If we use the approximation $e \approx 2.71828$, then
$2e + e^2 \approx 2(2.71828) + (2.71828)^2 \approx 5.43656 + 7.3890561984 \approx 12.8256161984$.

The exact value is $2e + e^2$.

Final Answer: The final answer is $\boxed{2e+e^2}$

In [7]:
# Get model's response
system_prompt = """You are a math agent. Respond with EXACTLY ONE of these formats:
1. FUNCTION_CALL: python_function_name|input
2. FINAL_ANSWER: [number]

where python_function_name is one of the followin:
1. strings_to_chars_to_int(string) It takes a word as input, and returns the ASCII INT values of characters in the word as a list
2. int_list_to_exponential_sum(list) It takes a list of integers and returns the sum of exponentials of those integers
3. fibonacci_numbers(int) It takes an integer, like 6, and returns first 6 integers in a fibonacci series as a list.
DO NOT include multiple responses. Give ONE response at a time."""

current_query= """Calculate the sum of exponentials of word "TSAI"""

prompt = f"{system_prompt}\n\nQuery: {current_query}"
response = client.models.generate_content(
    model="gemini-2.0-flash",
    contents=prompt
)

print(response.text)

FUNCTION_CALL: strings_to_chars_to_int|TSAI



In [8]:
import math

def strings_to_chars_to_int(string):
    return [ord(char) for char in string]

def int_list_to_exponential_sum(int_list):
    int_list = eval(int_list)
    return sum(math.exp(i) for i in int_list)

def fibonacci_numbers(n):
    if n <= 0:
        return []
    fib_sequence = [0, 1]
    for _ in range(2, n):
        fib_sequence.append(fib_sequence[-1] + fib_sequence[-2])
    return fib_sequence[:n]

In [9]:
response_text = response.text.strip()
response_text

'FUNCTION_CALL: strings_to_chars_to_int|TSAI'

In [10]:
_, function_info = response_text.split(":", 1)
_, function_info

('FUNCTION_CALL', ' strings_to_chars_to_int|TSAI')

In [11]:
func_name, params = [x.strip() for x in function_info.split("|", 1)]

func_name, params

('strings_to_chars_to_int', 'TSAI')

In [12]:
def function_caller(func_name, params):
    """Simple function caller that maps function names to actual functions"""
    function_map = {
        "strings_to_chars_to_int": strings_to_chars_to_int,
        "int_list_to_exponential_sum": int_list_to_exponential_sum,
        "fibonacci_numbers": fibonacci_numbers
    }
    
    if func_name in function_map:
        return function_map[func_name](params)
    else:
        return f"Function {func_name} not found"

In [13]:
iteration_result = function_caller(func_name, params)

In [14]:
# Get model's response
system_prompt = """You are a math agent solving problems in iterations. Respond with EXACTLY ONE of these formats:
1. FUNCTION_CALL: python_function_name|input
2. FINAL_ANSWER: [number]

where python_function_name is one of the followin:
1. strings_to_chars_to_int(string) It takes a word as input, and returns the ASCII INT values of characters in the word as a list
2. int_list_to_exponential_sum(list) It takes a list of integers and returns the sum of exponentials of those integers
3. fibonacci_numbers(int) It takes an integer, like 6, and returns first 6 integers in a fibonacci series as a list.
DO NOT include multiple responses. Give ONE response at a time."""

current_query= """Calculate the sum of exponentials of word "TSAI"""
iteration_1 = f"In the first iteration you called {func_name} with {params} parameters, and the function returned {iteration_result}. What should I do next?"

prompt = f"{system_prompt}\n\nQuery: {current_query}\n\n{iteration_1}"
response = client.models.generate_content(
    model="gemini-2.0-flash",
    contents=prompt
)

print(response.text)

FUNCTION_CALL: int_list_to_exponential_sum|[84, 83, 65, 73]



In [15]:
response_text = response.text.strip()
_, function_info = response_text.split(":", 1)
func_name, params = [x.strip() for x in function_info.split("|", 1)]
iteration_result = function_caller(func_name, params)
iteration_result

4.137991617878097e+36

In [16]:
# Get model's response
system_prompt = """You are a math agent solving problems in iterations. Respond with EXACTLY ONE of these formats:
1. FUNCTION_CALL: python_function_name|input
2. FINAL_ANSWER: [number]

where python_function_name is one of the followin:
1. strings_to_chars_to_int(string) It takes a word as input, and returns the ASCII INT values of characters in the word as a list
2. int_list_to_exponential_sum(list) It takes a list of integers and returns the sum of exponentials of those integers
3. fibonacci_numbers(int) It takes an integer, like 6, and returns first 6 integers in a fibonacci series as a list.
DO NOT include multiple responses. Give ONE response at a time."""

current_query= """Calculate the sum of exponentials of word "TSAI"""
iteration_1 = f"In the first iteration you called strings_to_chars_to_int with TSAI parameters, and the function returned {iteration_result}. What should I do next?"
iteration_2 = f"In the first iteration you called {func_name} with {params} parameters, and the function returned {iteration_result}. What should I do next?"
prompt = f"{system_prompt}\n\nQuery: {current_query}\n\n{iteration_1}\n\n{iteration_2}"
response = client.models.generate_content(
    model="gemini-2.0-flash",
    contents=prompt
)

print(response.text)

FINAL_ANSWER: 4.137991617878097e+36



In [17]:
max_iterations = 3
last_response = None
iteration = 0
iteration_response = []

system_prompt = """You are a math agent solving problems in iterations. Respond with EXACTLY ONE of these formats:
1. FUNCTION_CALL: python_function_name|input
2. FINAL_ANSWER: [number]

where python_function_name is one of the followin:
1. strings_to_chars_to_int(string) It takes a word as input, and returns the ASCII INT values of characters in the word as a list
2. int_list_to_exponential_sum(list) It takes a list of integers and returns the sum of exponentials of those integers
3. fibonacci_numbers(int) It takes an integer, like 6, and returns first 6 integers in a fibonacci series as a list.
DO NOT include multiple responses. Give ONE response at a time."""

query= """Calculate the sum of exponentials of word "TSAI"""

while iteration < max_iterations:
    print(f"\n--- Iteration {iteration + 1} ---")
    if last_response == None:
        current_query = query
    else:
        current_query = current_query + "\n\n" + " ".join(iteration_response)
        current_query = current_query + "  What should I do next?"

    # Get model's response
    prompt = f"{system_prompt}\n\nQuery: {current_query}"
    response = client.models.generate_content(
        model="gemini-2.0-flash",
        contents=prompt
    )
    
    response_text = response.text.strip()
    print(f"LLM Response: {response_text}")

    
    if response_text.startswith("FUNCTION_CALL:"):
        response_text = response.text.strip()
        _, function_info = response_text.split(":", 1)
        func_name, params = [x.strip() for x in function_info.split("|", 1)]
        iteration_result = function_caller(func_name, params)

    # Check if it's the final answer
    elif response_text.startswith("FINAL_ANSWER:"):
        print("\n=== Agent Execution Complete ===")
        break
        

    print(f"  Result: {iteration_result}")
    last_response = iteration_result
    iteration_response.append(f"In the {iteration + 1} iteration you called {func_name} with {params} parameters, and the function returned {iteration_result}.")

    iteration += 1
    
    



--- Iteration 1 ---
LLM Response: FUNCTION_CALL: strings_to_chars_to_int|TSAI
  Result: [84, 83, 65, 73]

--- Iteration 2 ---
LLM Response: FUNCTION_CALL: int_list_to_exponential_sum|[84, 83, 65, 73]
  Result: 4.137991617878097e+36

--- Iteration 3 ---
LLM Response: FINAL_ANSWER: [4.137991617878097e+36]

=== Agent Execution Complete ===
